Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java

```diff
@@ -75,15 +77,28 @@ public DartServerCompletionContributor() {
              protected void addCompletions(@NotNull final CompletionParameters parameters,
                                            @NotNull final ProcessingContext context,
                                            @NotNull final CompletionResultSet originalResultSet) {
-               VirtualFile file = DartResolveUtil.getRealVirtualFile(parameters.getOriginalFile());
+               final PsiFile originalFile = parameters.getOriginalFile();
+               final Project project = originalFile.getProject();
+
+               final CompletionSorter sorter = createSorter(parameters, originalResultSet.getPrefixMatcher());
+               final String uriPrefix = getPrefixIfCompletingUri(parameters);
+               final CompletionResultSet resultSet = uriPrefix != null
+                                                     ? originalResultSet.withRelevanceSorter(sorter).withPrefixMatcher(uriPrefix)
+                                                     : originalResultSet.withRelevanceSorter(sorter);
+
+               // If completion is requested in "Evaluate Expression" dialog, use runtime completion.
+               if (originalFile.getContext() != null) {
```

Comment:
`originalFile.getContext() != null` does not necessarily mean that we are inside Evaluate Expression window. There are many other cases when it happens. For example when editing a breakpoint condition.

Reply:
OK, it is probably misnomer then, but the feature is still useful, I think?
It seems to work in breakpoint conditions too.

We will not able to give runtime types of variables and expressions, but it is OK, they are designed to be optional - if we send empty lists for them.

Reply:
For some runtime expressions we can still do a good job even when not paused. See  https://github.com/JetBrains/intellij-plugins/pull/583 for a related CL to enable evaluating expressions when not paused at a breakpoint.

Reply:
Yes, when Alexander told me that context is set also for breakpoint conditions, I tried and can get good completion there too.

Reply:
`if (originalFile instanceof DartExpressionCodeFragment)` looks safer to me. Do you mind if I change it this way?

Reply:
Sounds good to me.